### PR TITLE
add routers

### PR DESCRIPTION
**Pull request by rayzhang-xingyunzh**
# Conflicts:
# app.js
# controllers/demoController.js
# package.json
# routes/router.js

### CODE REVIEW

File: controllers/userController.js [hunk 1]
```diff
@@ -1,5 +1,177 @@
...
 exports.loginByEmail = function(req,res){
+	var email = null;
+	var password = null;
+
+	const STATE_CHECK_PARAM = 1;
+	const STATE_LOGIN = 2;
+
+	stateMachine(null,STATE_CHECK_PARAM);
+
+	function stateMachine(err,toState){
```
Comment (williamyu-xingyunzh) [inline controllers/userController.js]: [non-blocker comment] I don't see any benefit to introduce state machine here. Check param is very ordinary action before perform other functional logic.

File: controllers/userController.js [hunk 2]
```diff
@@ -1,5 +1,177 @@
...
+				case STATE_LOGIN_EMAIL:
+					email = req.body.email;
+					password = req.body.password;
+
+					uidHelper.loginByEmail(email,password,function(err,result){
+						loginResult = result;
+						stateMachine(err,STATE_SEND_RESPONSE);
+					});
+				break;
+				case STATE_IS_FRIST_TIME:
```
Comment (williamyu-xingyunzh) [inline controllers/userController.js:86]: Are the states below reachable?

File: models/asset.js [hunk 3]
```diff
@@ -4,11 +4,21 @@ var Schema = mongoose.Schema;
 var assetSchema = Schema({
 	name:String,
 
-	content:String,
+	category:String,
 
-	createDate:Date,
+	version:{
+		no:Number,
+		date:Date
```
Comment (williamyu-xingyunzh) [inline models/asset.js:11]: Version should be simple string. If you want to add a create_date info, you can have a separate field to store it.

File: models/idea.js [hunk 4]
```diff
@@ -4,19 +4,25 @@ var Schema = mongoose.Schema;
 var ideaSchema = Schema({
 	name:String,
 
-	content:String,
+	background:String,
 
-	creator:{
+	innovator:{
 		type:Schema.Type.ObjectId,
 		ref:'Player'
```
Comment (williamyu-xingyunzh) [inline models/idea.js]: Since player is removed, this ref should be changed as well.

File: models/sprint.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+var mongoose = require('mongoose');
+var Schema = mongoose.Schema;
+
+var sprintSchema = Schema({
+	startDate:Date,
+
+	endDate:Date,
+
+	backlog:[{
```
Comment (williamyu-xingyunzh) [inline models/sprint.js]: "backlog" is not accurate here. It is only for the prioritized set of all the stories.

File: models/sprint.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+		type:Schema.Type.ObjectId,
+		ref:'UserStory'
+	}],
+
+	task:{
+		type:Schema.Type.ObjectId,
+		ref:'Tasks'
```
Comment (williamyu-xingyunzh) [inline models/sprint.js]: Do you have this schema?  's' is typo here? Tasks should be grouped and under some specific story.

File: models/team.js [hunk 7]
```diff
@@ -4,7 +4,22 @@ var Schema = mongoose.Schema;
 var teamSchema = Schema({
```
Comment (williamyu-xingyunzh) [inline models/team.js:4]: where is the uid we discussed, which is to identify the inheritance of the team.

File: models/team.js [hunk 8]
```diff
@@ -4,7 +4,22 @@ var Schema = mongoose.Schema;
...
+	project:{
+		type:Schema.Type.ObjectId,
+		ref:'Project'
+	},
+
+	createdDate:Date,
+
+	deactiveDate:Date,
+
+	activeDate:Date
```
Comment (williamyu-xingyunzh) [inline models/team.js:20]: Do you need both create date and active date? What's the difference you think?

File: models/task.js [hunk 9]
```diff
@@ -4,14 +4,14 @@ var Schema = mongoose.Schema;
 var taskSchema = Schema({
 	description:String,
 
-	duration:Number,
+	duedate:Date,
 
 	assignee:{
 		type:Schema.Type.ObjectId,
 		ref:'TeamMember'
```
Comment (williamyu-xingyunzh) [inline models/task.js]: Why is it 'TeamMember', other than 'User' here?

File: routes/accessControl.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+var tokenHelper = require('../util/shared/tokenHelper.js');
+var util = require('../util/shared/util.js');
+
+
+module.exports = function(app){
```
Comment (williamyu-xingyunzh) [inline routes/accessControl.js]: [blocker]This is definitely wrong here: 
1. Controller should not do any routing chain changes. Move this logic to routes module.
2. You should not apply any route specific info to app directly. You should apply it to rootRouter instead. This code here breaks the contextRoot feature.

File: routes/accessControl.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+	app.use('/api/*', function(req, res, next) {
+		console.log('inside ac');
+
+		var tokenString = req.body.token;
```
Comment (williamyu-xingyunzh) [inline routes/accessControl.js]: [b]The token should be part of header of the request, not the body here.
You should not ask every request to be post and have a body field "token".

File: util/shared/tokenHelper.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
+
```
Comment (williamyu-xingyunzh) [inline util/shared/tokenHelper.js]: The token logic is definitely not just a "util" method. It should be in authentication module. Util is for just general method, like string, math, wrapper, pure convenient method. It should not contain any business logic. 
Especially secret is very important and specific here.

**Review by williamyu-xingyunzh — CHANGES_REQUESTED** [review]
Please clarify/reply or accept/implement all the comments in this review.

Apart from business questions/coments, here are 2 general comments for programing perspecitve:
1. Abuse of "State Machine" method.
2. Abuse of "util"

File: models/idea.js (hunk 4 above)
Comment (realmorrisliu) [reply models/idea.js]: Should the `Schema.Type.ObjectId` be `Schema.Types.ObjectId`?

File: controllers/userController.js (hunk 2 above)
Comment (rayzhang-xingyunzh) [reply controllers/userController.js:86]: fixed

File: models/asset.js (hunk 3 above)
Comment (rayzhang-xingyunzh) [reply models/asset.js:11]: 1. A string? not number?
2. a version date is the date for the specific version

File: models/idea.js (hunk 4 above)
Comment (rayzhang-xingyunzh) [reply models/idea.js]: fixed

File: models/sprint.js (hunk 5 above)
Comment (rayzhang-xingyunzh) [reply models/sprint.js]: fixed

File: models/sprint.js (hunk 6 above)
Comment (rayzhang-xingyunzh) [reply models/sprint.js]: It is a typo.fixed

File: models/team.js (hunk 8 above)
Comment (rayzhang-xingyunzh) [reply models/team.js:20]: I was thinking about the create date of the inherited team, but maybe we can use the earliest activeDate as the create date. How do you think @williamyu-xingyunzh ?

File: models/asset.js (hunk 3 above)
Comment (williamyu-xingyunzh) [reply models/asset.js:11]: yes, version should be a simple string as I mentioned. Don't make simple complex.

File: models/team.js (hunk 8 above)
Comment (williamyu-xingyunzh) [reply models/team.js:20]: Yeah, please ensure the data is not duplicated in your model.

File: controllers/userController.js (hunk 1 above)
Comment (rayzhang-xingyunzh) [reply controllers/userController.js]: fixed

File: models/team.js (hunk 7 above)
Comment (rayzhang-xingyunzh) [reply models/team.js:4]: teamId

File: routes/accessControl.js (hunk 10 above)
Comment (rayzhang-xingyunzh) [reply routes/accessControl.js]: fixed

File: models/task.js (hunk 9 above)
Comment (rayzhang-xingyunzh) [reply models/task.js]: fixed

File: models/comment.js [hunk 13]
```diff
@@ -3,7 +3,7 @@ var Schema = mongoose.Schema;
 
 var commentSchema = Schema({
 	user:{
-		type:Schema.type.ObjectId,
+		type:Schema.types.ObjectId,
```
Comment (williamyu-xingyunzh) [inline models/comment.js:6]: Capital “T”?

File: models/discussMinutes.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
+var mongoose = require('mongoose');
+var Schema = mongoose.Schema;
+
+var discussMinutesSchema = Schema({
+	creator:{
+		type:Schema.types.ObjectId,
```
Comment (williamyu-xingyunzh) [inline models/discussMinutes.js:6]: Types
Your editor did not work well to help you. You should make it work.

File: models/sprint.js [hunk 15]
```diff
@@ -6,18 +6,18 @@ var sprintSchema = Schema({
...
 	endDate:Date,
 
-	backlog:[{
-		type:Schema.Type.ObjectId,
+	userStoris:[{
+		type:Schema.Types.ObjectId,
 		ref:'UserStory'
 	}],
 
 	task:{
```
Comment (williamyu-xingyunzh) [inline models/sprint.js:14]: only one task allowed in one sprint?

File: models/task.js [hunk 16]
```diff
@@ -7,8 +7,13 @@ var taskSchema = Schema({
 	duedate:Date,
```
Comment (williamyu-xingyunzh) [inline models/task.js:7]: add effort field.

**Review by williamyu-xingyunzh — APPROVED** [review]
version in asset.js is not fixed.
There are stills lots of typo. 
Please unit test your code before writing more.  agile way of work is making a bit, testing the bit, then another bit. Do not do it all in all.

**Pull request merged**